File: src/main/java/de/esymetric/jerusalem/ownDataRepresentation/fileSystem/PartitionedWayCostFile.java
package de.esymetric.jerusalem.ownDataRepresentation.fileSystem;

import java.io.File;
import java.io.IOException;

import de.esymetric.jerusalem.ownDataRepresentation.Transition;
import de.esymetric.jerusalem.utils.BufferedRandomAccessFile;
import de.esymetric.jerusalem.utils.BufferedRandomAccessFileCache;

public class PartitionedWayCostFile {

	final static long SENTENCE_LENGTH = 24L;
	final static String FILENAME = "wayCost.data";

	String dataDirectoryPath, filePath;
	boolean readOnly;
	LatLonDir currentLatLonDir = new LatLonDir(-1000, -1000);

	BufferedRandomAccessFile raf;
	BufferedRandomAccessFileCache rafCache = new BufferedRandomAccessFileCache();

	int numberOfWayCosts;

	public PartitionedWayCostFile(String dataDirectoryPath, boolean readOnly) {
		this.dataDirectoryPath = dataDirectoryPath;
		this.readOnly = readOnly;
		if( readOnly )
			rafCache.setMaxCacheSize(30);
		else
			rafCache.setMaxCacheSize(10);
	}

	void checkAndCreateRandomAccessFile(LatLonDir lld) {
		//LatLonDir newLatLonDir = new LatLonDir(lat, lng);
		if (lld.equals(currentLatLonDir))
			return;

		currentLatLonDir = lld;

		filePath = currentLatLonDir.makeDir(dataDirectoryPath, !readOnly)
				+ File.separatorChar + FILENAME;
		raf = rafCache.getRandomAccessFile(filePath, readOnly);
		try {
			long fileLength = raf.length();
			numberOfWayCosts = (int) (fileLength / SENTENCE_LENGTH);
		} catch (IOException e) {
			e.printStackTrace();
		}

	}

	public void close() {
		rafCache.close();
	}





	public int insertWay(LatLonDir lld, double costFoot, double costBike,
			double costRacingBike, double costMountainBike, double costCar,
			double costCarShortest) {
		checkAndCreateRandomAccessFile(lld);

		try {
			int id = numberOfWayCosts;

			raf.seek((long)numberOfWayCosts * SENTENCE_LENGTH);

			raf.writeFloat((float) costFoot);
			raf.writeFloat((float) costBike);
			raf.writeFloat((float) costRacingBike);
			raf.writeFloat((float) costMountainBike);
			raf.writeFloat((float) costCar);
			raf.writeFloat((float) costCarShortest);
			numberOfWayCosts++;
			return id;
		} catch (IOException e) {
			e.printStackTrace();
			return -1;
		}
	}

	public boolean readTransitionCost(LatLonDir lld, int wayCostID, Transition t) {
		checkAndCreateRandomAccessFile(lld);

		try {
			raf.seek((long) wayCostID * SENTENCE_LENGTH);
			t.costFoot = raf.readFloat();
			t.costBike = raf.readFloat();
			t.costRacingBike = raf.readFloat();
			t.costMountainBike = raf.readFloat();
			t.costCar = raf.readFloat();
			t.costCarShortest = raf.readFloat();
			return true;
		} catch (IOException e) {
			e.printStackTrace();
			return false;
		}
	}
	public void deleteAllWayCostFiles() {
		for(File f : new File(dataDirectoryPath).listFiles() )
			if( f.isDirectory() && f.getName().startsWith("lat_"))
				for(File g : f.listFiles() )
					if( g.isDirectory() && g.getName().startsWith("lng_"))
						for(File h : g.listFiles() )
							if( h.isFile() && h.getName().equals(FILENAME) )
								h.delete();

	}
}

File: src/main/java/de/esymetric/jerusalem/ownDataRepresentation/fileSystem/PartitionedWayCostFile.kt
package de.esymetric.jerusalem.ownDataRepresentation.fileSystem

import de.esymetric.jerusalem.ownDataRepresentation.Transition
import de.esymetric.jerusalem.routing.RoutingHeuristics.Companion.BLOCKED_WAY_COST
import de.esymetric.jerusalem.utils.BufferedRandomAccessFile
import de.esymetric.jerusalem.utils.BufferedRandomAccessFileCache
import java.io.File
import java.io.IOException

class PartitionedWayCostFile(var dataDirectoryPath: String, var readOnly: Boolean) {
    var filePath: String? = null
    var currentLatLonDir = LatLonDir(-1000.0, -1000.0)
    var raf: BufferedRandomAccessFile? = null
    private var rafCache = BufferedRandomAccessFileCache()
    private var numberOfWayCosts = 0

    init {
        if (readOnly) rafCache.setMaxCacheSize(30) else rafCache.setMaxCacheSize(10)
    }

    private fun checkAndCreateRandomAccessFile(lld: LatLonDir) {
        if (lld.equals(currentLatLonDir)) return
        currentLatLonDir = lld
        filePath = (currentLatLonDir.makeDir(dataDirectoryPath, !readOnly)
                + File.separatorChar + FILENAME)
        raf = rafCache.getRandomAccessFile(filePath!!, readOnly)
        try {
            val fileLength = raf!!.length()
            numberOfWayCosts = (fileLength / SENTENCE_LENGTH).toInt()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun close() {
        rafCache.close()
    }

    fun insertWay(
        lld: LatLonDir, costFoot: Double, costBike: Double,
        costRacingBike: Double, costMountainBike: Double, costCar: Double,
        costCarShortest: Double
    ): Int {
        checkAndCreateRandomAccessFile(lld)
        return try {
            val id = numberOfWayCosts
            raf!!.seek(numberOfWayCosts.toLong() * SENTENCE_LENGTH)
            raf!!.writeUShort(costDouble2Short(costFoot))
            raf!!.writeUShort(costDouble2Short(costBike))
            raf!!.writeUShort(costDouble2Short(costRacingBike))
            raf!!.writeUShort(costDouble2Short(costMountainBike))
            raf!!.writeUShort(costDouble2Short(costCar))
            raf!!.writeUShort(costDouble2Short(costCarShortest))
            numberOfWayCosts++
            id
        } catch (e: IOException) {
            e.printStackTrace()
            -1
        }
    }

    fun readTransitionCost(lld: LatLonDir, wayCostID: Int, t: Transition): Boolean {
        checkAndCreateRandomAccessFile(lld)
        return try {
            raf!!.seek(wayCostID.toLong() * SENTENCE_LENGTH)
            t.costFoot = short2DoubleCost(raf!!.readUShort())
            t.costBike = short2DoubleCost(raf!!.readUShort())
            t.costRacingBike = short2DoubleCost(raf!!.readUShort())
            t.costMountainBike = short2DoubleCost(raf!!.readUShort())
            t.costCar = short2DoubleCost(raf!!.readUShort())
            t.costCarShortest = short2DoubleCost(raf!!.readUShort())
            true
        } catch (e: IOException) {
            e.printStackTrace()
            false
        }
    }

    fun deleteAllWayCostFiles() {
        for (f in File(dataDirectoryPath).listFiles()) if (f.isDirectory && f.name.startsWith("lat_")) for (g in f.listFiles()) if (g.isDirectory && g.name.startsWith(
                "lng_"
            )
        ) for (h in g.listFiles()) if (h.isFile && h.name == FILENAME) h.delete()
    }

    private fun costDouble2Short(cost: Double) : UShort {
        if (cost == BLOCKED_WAY_COST) return BLOCKED_WAY_COST_USHORT

        var costM = cost * MULT_FACTOR_SHORT_TO_DOUBLE
        if (costM > 64000.0 || costM < 0.0) {
            println("Error: invalid cost " + cost)
            costM = 64000.0
        }
        return costM.toUInt().toUShort()
    }

    private fun short2DoubleCost(cost: UShort): Double {
        if ( cost == BLOCKED_WAY_COST_USHORT ) return BLOCKED_WAY_COST

        return cost.toDouble() / MULT_FACTOR_SHORT_TO_DOUBLE
    }

    companion object {
        const val SENTENCE_LENGTH = 12L
        const val FILENAME = "wayCost.data"
        const val MULT_FACTOR_SHORT_TO_DOUBLE = 64.0
        const val BLOCKED_WAY_COST_USHORT = UShort.MAX_VALUE
    }
}